### PR TITLE
[CodeCompletion] Escape declaration base name if needed

### PR DESCRIPTION
**Pull request by rintaro**
These should be escaped identifiers in code completion:
- As primary expression: Any keyword name except for `self` and `Self`.
- After dot: Something named `` `init` ``.

rdar://problem/16232627

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -1823,6 +1823,27 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+    auto NameStr = Name.userFacingName();
+    bool shouldEscapeKeywords =
+        // Special names (i.e. 'init') are always displayed as its user facing
+        // name.
+        !Name.isSpecial() &&
+        // User can write any keyword after '.' exept for non-special `init`.
+        // e.g. 'func `init`()' should be 'expe.`init`()'.
+        (!ExprType || NameStr == "init") &&
+        // 'self' and 'Self' are special
+        NameStr != "self" && NameStr != "Self";
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: I didn't see any error writing x.Self as a property reference.

**Review by benlangmuir — COMMENTED** [review]
Mostly LGTM

File: lib/IDE/CodeCompletion.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: `foo.self`, `foo.Self`, `self` and `Self` are all valid. We don't have to escape `self` and `Self` .
```swift
class C {
  func foo() {
    var `self` = 12
    let `Self` = 42
    self += Self
    print(self)
  }
}
```

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: Ah, I was wrong.
```swift
class C {
   var `self` = 12
}
print(C().self) // C
print(C().`self`) // 12
```
I will update this part to clarify the intent.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test OS X platform

**Comment by rintaro** [comment]
@swift-ci Please smoke test OS X platform

**Comment by rintaro** [comment]
@swift-ci Please smoke test OS X platform

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**